File: src/1. Layers/1.3 Domain/Well.Domain/ValueObjects/Delivery.cs
namespace PH.Well.Domain.ValueObjects
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using Common.Extensions;
    using Enums;

    public class Delivery
    {
        public Delivery()
        {
            Lines = new List<DeliveryLine>();
        }

        public int Id { get; set; } 

        public IList<DeliveryLine> Lines { get; set; }

        public string RouteNumber { get; set; }

        public DateTime RouteDate { get; set; }

        public int DropId { get; set; }

        public string InvoiceNumber { get; set; }

        public string AccountCode { get; set; }

        public string AccountName { get; set; }

        public string JobStatus { get; set; }

        public DateTime DeliveryDate { get; set; }

        public string FormattedDeliveryDate => this.DeliveryDate.ToShortDateString();

        public string Reason { get; set; }

        public string Action { get; set; }

        public string Assigned { get; set; }

        public string AccountId { get; set; }

        public int BranchId { get; set; }

        public string IdentityName { get; set; }
        public string CurrentUserIdentity { get; set; }

        public string CashOnDelivery { get; set; }
        public bool IsCashOnDelivery => string.IsNullOrWhiteSpace(CashOnDelivery) == false;

        public bool CanAction => string.Equals(CurrentUserIdentity, IdentityName, StringComparison.OrdinalIgnoreCase) &&
                                 JobStatus != Enums.JobStatus.Resolved.ToString();

        public bool CanSubmit => CanAction && Lines.All(l => l.CanSubmit);

        public string TotalCredit { get; set; }

        public string PendingCreditCreatedBy { get; set; }

        public string FormattedPendingCreditCreatedBy {
            get
            {
                if (!string.IsNullOrWhiteSpace(this.PendingCreditCreatedBy) && this.PendingCreditCreatedBy.Contains("\\"))
                {
                    return this.PendingCreditCreatedBy.Split('\\')[1];
                }

                return this.PendingCreditCreatedBy;
            }
        }

        public decimal TotalCreditValueForThreshold => Lines.Sum(d => d.CreditValueForThreshold());

        public ThresholdLevel? ThresholdLevel { get; set; }

        public decimal ThresholdAmount { get; set; }

        public string CreditThresholdLevel
            => ThresholdLevel != null ? Enum<ThresholdLevel>.GetDescription(ThresholdLevel) : "";

        public bool ThresholdLevelValid => this.ThresholdAmount >= this.TotalCreditValueForThreshold;

        public int TotalOutersShort { get; set; }

        public bool IsPendingCredit { get; set; }

        public ProofOfDelivery? ProofOfDelivery { get; set; }
        public bool IsProofOfDelivery => ProofOfDelivery.HasValue;

        public bool IsAssignedTo(string identityName)
        {
            return string.Equals(IdentityName, identityName, StringComparison.OrdinalIgnoreCase);
        }

        public bool IsOutstanding => DeliveryDate.Date < DateTime.Today.Date;

        public bool OuterDiscrepancyFound { get; set; }

        public int ToBeAdvisedCount => OuterDiscrepancyFound ? TotalOutersShort : 0;

        public bool CanBulkCredit => Lines.All(l => l.HasNoActions);
    }
}

File: src/1. Layers/1.1 Presentation/Well.Dashboard/app/delivery/model/delivery.ts
import { DeliveryLine } from './deliveryLine';

export class Delivery
{
    constructor(delivery: Delivery)
    {
        if (delivery)
        {
            this.id = delivery.id;
            this.accountCode = delivery.accountCode;
            this.outerCount = delivery.outerCount;
            this.outerDiscrepancyFound = delivery.outerDiscrepancyFound;
            this.totalOutersShort = delivery.totalOutersShort;
            this.accountName = delivery.accountName;
            this.accountAddress = delivery.accountAddress;
            this.invoiceNumber = delivery.invoiceNumber;
            this.contactName = delivery.contactName;
            this.phoneNumber = delivery.phoneNumber;
            this.mobileNumber = delivery.mobileNumber;
            this.jobStatus = delivery.jobStatus;
            this.cashOnDelivery = delivery.cashOnDelivery;
            this.isCashOnDelivery = delivery.isCashOnDelivery;
            this.canAction = delivery.canAction;
            this.canSubmit = delivery.canSubmit;
            this.isPendingCredit = delivery.isPendingCredit;
            this.grnNumber = delivery.grnNumber;
            this.branchId = delivery.branchId;
            this.grnProcessType = delivery.grnProcessType;
            this.proofOfDelivery = delivery.proofOfDelivery;
            this.isProofOfDelivery = delivery.isProofOfDelivery;
            this.canBulkCredit = delivery.canBulkCredit;

            if (delivery.exceptionDeliveryLines)
            {
                delivery.jobStatus = 'Exception';
                for (const line of delivery.exceptionDeliveryLines)
                {
                    this.exceptionDeliveryLines.push(new DeliveryLine(line));
                }
            }

            if (delivery.cleanDeliveryLines)
            {
                delivery.jobStatus = 'Clean';
                for (const line of delivery.cleanDeliveryLines)
                {
                    this.cleanDeliveryLines.push(new DeliveryLine(line));
                }
            }
        }
    }
    public id: number;
    public accountCode: string;
    public outerCount: number;  
    public outerDiscrepancyFound: boolean;
    public totalOutersShort: number;
    public accountName: string;
    public accountAddress: string;
    public invoiceNumber: string;
    public contactName: string;
    public phoneNumber: string;
    public mobileNumber: string;
    public jobStatus: string;
    public cashOnDelivery: string;
    public isCashOnDelivery: boolean;
    public canAction: boolean;
    public canSubmit: boolean;
    public isPendingCredit: boolean;
    public grnNumber: string;
    public exceptionDeliveryLines: DeliveryLine[] = new Array<DeliveryLine>();
    public branchId: number;
    public grnProcessType: number;
    public proofOfDelivery: number;
    public isProofOfDelivery: boolean;
    public cleanDeliveryLines: DeliveryLine[] = new Array<DeliveryLine>();
    public canBulkCredit: boolean;

    public isCleanOnInit(): boolean
    {
        let clean = true;

        for (const line of this.exceptionDeliveryLines)
        {
            if (line.isCleanOnInit === false)
            {
                clean = false;
            }
        }

        for (const line of this.cleanDeliveryLines)
        {
            if (line.isCleanOnInit === false)
            {
                clean = false;
            }
        }

        return clean;
    };

    public isClean(): boolean
    {
        let clean = true;

        for (const line of this.exceptionDeliveryLines)
        {
            if (line.isClean() === false)
            {
                clean = false;
            }
        }

        for (const line of this.cleanDeliveryLines)
        {
            if (line.isClean() === false)
            {
                clean = false;
            }
        }

        return clean;
    }
}